### PR TITLE
Refactor Model to Pytorch Lightning style

### PR DESCRIPTION
**Pull request by henryyang42**
- Removed evaluate.py as we used pl.Trainer flow now
- Parameter search follows the new training flow

Note
- Should we call get_metric_dict before __repr__ in MultiLabelMetrics?

### CODE REVIEW

**Comment by henryyang42** [comment]
@Eleven1Liu I have test new training flow on parameter search using this command ``python3 search_params.py --config example_config/rcv1/cnn_tune.yml --search_alg basic_variant``. Please help evaluate if this is expected behavior and let us know if weird thing happened 

```
Number of trials: 50/50 (50 TERMINATED)
+-----------------------+------------+-------+-----------+-----------------+-----------+-----------+-----------+------------+------------+------------+
| Trial name            | status     | loc   |   dropout |   learning_rate |   val_P@1 |   val_P@3 |   val_P@5 |   test_P@1 |   test_P@3 |   test_P@5 |
|-----------------------+------------+-------+-----------+-----------------+-----------+-----------+-----------+------------+------------+------------|
| Trainable_7e3d9_00000 | TERMINATED |       |       0.2 |     0.000483417 |  0.968251 |  0.797048 |  0.553305 |   0.968251 |   0.797048 |   0.553305 |
| Trainable_7e3d9_00001 | TERMINATED |       |       0.4 |     0.000448811 |  0.964579 |  0.791505 |  0.549158 |   0.964579 |   0.791505 |   0.549158 |
| Trainable_7e3d9_00002 | TERMINATED |       |       0.6 |     0.000621522 |  0.963067 |  0.798272 |  0.549374 |   0.963067 |   0.798272 |   0.549374 |
| Trainable_7e3d9_00003 | TERMINATED |       |       0.2 |     0.000464772 |  0.968898 |  0.802736 |  0.556328 |   0.968898 |   0.802736 |   0.556328 |
| Trainable_7e3d9_00004 | TERMINATED |       |       0.6 |     0.00048336  |  0.965227 |  0.800432 |  0.552484 |   0.965227 |   0.800432 |   0.552484 |
| Trainable_7e3d9_00005 | TERMINATED |       |       0.2 |     0.000489337 |  0.969762 |  0.801728 |  0.555076 |   0.969762 |   0.801728 |   0.555076 |
| Trainable_7e3d9_00006 | TERMINATED |       |       0.4 |     0.000380692 |  0.969114 |  0.806552 |  0.554989 |   0.969114 |   0.806552 |   0.554989 |
| Trainable_7e3d9_00007 | TERMINATED |       |       0.6 |     0.000366603 |  0.961123 |  0.803312 |  0.553521 |   0.961123 |   0.803312 |   0.553521 |
| Trainable_7e3d9_00008 | TERMINATED |       |       0.2 |     0.000387541 |  0.969546 |  0.806047 |  0.556674 |   0.969546 |   0.806047 |   0.556674 |
| Trainable_7e3d9_00009 | TERMINATED |       |       0.2 |     0.000411964 |  0.967171 |  0.801584 |  0.555378 |   0.967171 |   0.801584 |   0.555378 |
| Trainable_7e3d9_00010 | TERMINATED |       |       0.6 |     0.000852691 |  0.960259 |  0.786465 |  0.543629 |   0.960259 |   0.786465 |   0.543629 |
| Trainable_7e3d9_00011 | TERMINATED |       |       0.8 |     0.00030374  |  0.951836 |  0.796904 |  0.549287 |   0.951836 |   0.796904 |   0.549287 |
| Trainable_7e3d9_00012 | TERMINATED |       |       0.8 |     0.000591273 |  0.953996 |  0.791361 |  0.54622  |   0.953996 |   0.791361 |   0.54622  |
| Trainable_7e3d9_00013 | TERMINATED |       |       0.2 |     0.00066435  |  0.967603 |  0.795608 |  0.553521 |   0.967603 |   0.795608 |   0.553521 |
| Trainable_7e3d9_00014 | TERMINATED |       |       0.6 |     0.000431466 |  0.965875 |  0.807127 |  0.558402 |   0.965875 |   0.807127 |   0.558402 |
| Trainable_7e3d9_00015 | TERMINATED |       |       0.2 |     0.0009045   |  0.968035 |  0.796688 |  0.55041  |   0.968035 |   0.796688 |   0.55041  |
| Trainable_7e3d9_00016 | TERMINATED |       |       0.2 |     0.000649754 |  0.965875 |  0.802016 |  0.556501 |   0.965875 |   0.802016 |   0.556501 |
| Trainable_7e3d9_00017 | TERMINATED |       |       0.2 |     0.000398598 |  0.970842 |  0.804248 |  0.55892  |   0.970842 |   0.804248 |   0.55892  |
| Trainable_7e3d9_00018 | TERMINATED |       |       0.6 |     0.000439437 |  0.961771 |  0.795392 |  0.550238 |   0.961771 |   0.795392 |   0.550238 |
| Trainable_7e3d9_00019 | TERMINATED |       |       0.6 |     0.000949885 |  0.961771 |  0.797048 |  0.551102 |   0.961771 |   0.797048 |   0.551102 |
| Trainable_7e3d9_00020 | TERMINATED |       |       0.4 |     0.000429676 |  0.968035 |  0.8018   |  0.555076 |   0.968035 |   0.8018   |   0.555076 |
| Trainable_7e3d9_00021 | TERMINATED |       |       0.2 |     0.0009167   |  0.968898 |  0.795752 |  0.551577 |   0.968898 |   0.795752 |   0.551577 |
| Trainable_7e3d9_00022 | TERMINATED |       |       0.6 |     0.000652805 |  0.965011 |  0.798632 |  0.554557 |   0.965011 |   0.798632 |   0.554557 |
| Trainable_7e3d9_00023 | TERMINATED |       |       0.2 |     0.000785091 |  0.968251 |  0.794744 |  0.550497 |   0.968251 |   0.794744 |   0.550497 |
| Trainable_7e3d9_00024 | TERMINATED |       |       0.8 |     0.000812183 |  0.949244 |  0.794025 |  0.549546 |   0.949244 |   0.794025 |   0.549546 |
| Trainable_7e3d9_00025 | TERMINATED |       |       0.8 |     0.000765568 |  0.949028 |  0.788913 |  0.545572 |   0.949028 |   0.788913 |   0.545572 |
| Trainable_7e3d9_00026 | TERMINATED |       |       0.4 |     0.000388644 |  0.969546 |  0.80252  |  0.553866 |   0.969546 |   0.80252  |   0.553866 |
| Trainable_7e3d9_00027 | TERMINATED |       |       0.4 |     0.000741749 |  0.966091 |  0.806623 |  0.55676  |   0.966091 |   0.806623 |   0.55676  |
| Trainable_7e3d9_00028 | TERMINATED |       |       0.2 |     0.000302031 |  0.969762 |  0.802664 |  0.556803 |   0.969762 |   0.802664 |   0.556803 |
| Trainable_7e3d9_00029 | TERMINATED |       |       0.2 |     0.000893627 |  0.968467 |  0.801584 |  0.55473  |   0.968467 |   0.801584 |   0.55473  |
| Trainable_7e3d9_00030 | TERMINATED |       |       0.4 |     0.000852535 |  0.966091 |  0.803168 |  0.554816 |   0.966091 |   0.803168 |   0.554816 |
| Trainable_7e3d9_00031 | TERMINATED |       |       0.4 |     0.000337058 |  0.968467 |  0.80468  |  0.556587 |   0.968467 |   0.80468  |   0.556587 |
| Trainable_7e3d9_00032 | TERMINATED |       |       0.4 |     0.00064386  |  0.966739 |  0.806911 |  0.556587 |   0.966739 |   0.806911 |   0.556587 |
| Trainable_7e3d9_00033 | TERMINATED |       |       0.2 |     0.00082434  |  0.969978 |  0.802088 |  0.554082 |   0.969978 |   0.802088 |   0.554082 |
| Trainable_7e3d9_00034 | TERMINATED |       |       0.8 |     0.000727166 |  0.948596 |  0.793017 |  0.548121 |   0.948596 |   0.793017 |   0.548121 |
| Trainable_7e3d9_00035 | TERMINATED |       |       0.6 |     0.000886566 |  0.964363 |  0.803888 |  0.553953 |   0.964363 |   0.803888 |   0.553953 |
| Trainable_7e3d9_00036 | TERMINATED |       |       0.2 |     0.000478307 |  0.968683 |  0.804536 |  0.55676  |   0.968683 |   0.804536 |   0.55676  |
| Trainable_7e3d9_00037 | TERMINATED |       |       0.2 |     0.000512601 |  0.970842 |  0.807271 |  0.555551 |   0.970842 |   0.807271 |   0.555551 |
| Trainable_7e3d9_00038 | TERMINATED |       |       0.4 |     0.000957288 |  0.966091 |  0.791505 |  0.551318 |   0.966091 |   0.791505 |   0.551318 |
| Trainable_7e3d9_00039 | TERMINATED |       |       0.2 |     0.000918363 |  0.967819 |  0.802304 |  0.554773 |   0.967819 |   0.802304 |   0.554773 |
| Trainable_7e3d9_00040 | TERMINATED |       |       0.6 |     0.000538655 |  0.961339 |  0.800792 |  0.552138 |   0.961339 |   0.800792 |   0.552138 |
| Trainable_7e3d9_00041 | TERMINATED |       |       0.6 |     0.000402684 |  0.966307 |  0.80252  |  0.555032 |   0.966307 |   0.80252  |   0.555032 |
| Trainable_7e3d9_00042 | TERMINATED |       |       0.6 |     0.000831967 |  0.964795 |  0.805184 |  0.557019 |   0.964795 |   0.805184 |   0.557019 |
| Trainable_7e3d9_00043 | TERMINATED |       |       0.4 |     0.00067683  |  0.967603 |  0.804392 |  0.556026 |   0.967603 |   0.804392 |   0.556026 |
| Trainable_7e3d9_00044 | TERMINATED |       |       0.8 |     0.000766525 |  0.949892 |  0.793737 |  0.548121 |   0.949892 |   0.793737 |   0.548121 |
| Trainable_7e3d9_00045 | TERMINATED |       |       0.4 |     0.00076713  |  0.966091 |  0.805184 |  0.556847 |   0.966091 |   0.805184 |   0.556847 |
| Trainable_7e3d9_00046 | TERMINATED |       |       0.8 |     0.000937405 |  0.948164 |  0.781713 |  0.542333 |   0.948164 |   0.781713 |   0.542333 |
| Trainable_7e3d9_00047 | TERMINATED |       |       0.8 |     0.000912319 |  0.94838  |  0.784305 |  0.540216 |   0.94838  |   0.784305 |   0.540216 |
| Trainable_7e3d9_00048 | TERMINATED |       |       0.4 |     0.000485815 |  0.967387 |  0.806911 |  0.556933 |   0.967387 |   0.806911 |   0.556933 |
| Trainable_7e3d9_00049 | TERMINATED |       |       0.2 |     0.000986513 |  0.96933  |  0.803096 |  0.555378 |   0.96933  |   0.803096 |   0.555378 |
+-----------------------+------------+-------+-----------+-----------------+-----------+-----------+-----------+------------+------------+------------+


2021-06-13 07:21:48,382	INFO tune.py:549 -- Total run time: 4519.44 seconds (4519.26 seconds for the tuning loop).
2021-06-13 07:21:51,669 INFO:
Wall time: 4597.562280893326
```

File: libmultilabel/model.py [hunk 1]
```diff
@@ -1,218 +1,108 @@
...
-                                       momentum=self.config.momentum,
-                                       weight_decay=self.config.weight_decay)
+            optimizer = optim.SGD(parameters, self.config.learning_rate,
+                                  momentum=self.config.momentum,
+                                  weight_decay=self.config.weight_decay)
         elif optimizer_name == 'adam':
-            self.optimizer = optim.Adam(parameters, weight_decay=self.config.weight_decay, lr=self.config.learning_rate)
-
+            optimizer = optim.Adam(
+                parameters, weight_decay=self.config.weight_decay, lr=self.config.learning_rate)
```
Comment (cjlin1) [inline libmultilabel/model.py:36]: the way to split things in different lines different from SGD..

File: libmultilabel/model.py [hunk 2]
```diff
@@ -1,218 +1,108 @@
...
+        elif optimizer_name == 'adamw':
+            optimizer = optim.AdamW(
+                parameters, weight_decay=self.config.weight_decay, lr=self.config.learning_rate)
         else:
-            raise RuntimeError('Unsupported optimizer: %s' % self.config.optimizer)
+            raise RuntimeError('Unsupported optimizer: %s' %
```
Comment (cjlin1) [inline libmultilabel/model.py:41]: why this change?

Comment (henryyang42) [reply libmultilabel/model.py:41]: Both change are caused by editor auto-formatting. Fixed.

File: search_params.py [hunk 3]
```diff
@@ -36,26 +38,53 @@ def step(self):
         )
         logging.info(f'Run name: {self.config.run_name}')
 
+        checkpoint_dir = os.path.join(
+            self.config.result_dir, self.config.run_name)
+        best_checkpoint_path = os.path.join(checkpoint_dir, 'best_model.ckpt')
+        last_checkpoint_path = os.path.join(
+            checkpoint_dir, 'lastest_model.ckpt')
```
Comment (Eleven1Liu) [inline search_params.py:45]: Why we use `.ckpt` instead of `.pt` here? Is this a fix type of `load_from_checkpoint`?

**Comment by Eleven1Liu** [comment]
For CAML, we should also consider the situation when `config.init_weight` is `null`.  Please take a look at [this file](https://github.com/ASUS-AICS/LibMultiLabel/blob/e62a49567e1f56f88c29b5a370ff8a57315f36cf/libmultilabel/networks/__init__.py#L11).

File: search_params.py [hunk 4]
```diff
@@ -4,24 +4,26 @@
...
 
-logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)s:%(message)s')
+logging.basicConfig(level=logging.INFO,
+                    format='%(asctime)s %(levelname)s:%(message)s')
 
 
 class Trainable(tune.Trainable):
     def setup(self, config, data):
-        self.config = ArgDict(config)
+        self.config = AttributeDict(config)
```
Comment (Eleven1Liu) [inline search_params.py:26]: The second `set_seed()` cannot be ignored since we want to ensure the reproducibility. I think merge master would help?

File: search_params.py [hunk 5]
```diff
@@ -125,13 +154,18 @@ def main():
     parser = argparse.ArgumentParser()
     parser.add_argument(
         '--config', help='Path to configuration file (default: %(default)s). Please specify a config with all arguments in LibMultiLabel/main.py::get_config.')
-    parser.add_argument('--cpu_count', type=int, default=4, help='Number of CPU per trial (default: %(default)s)')
-    parser.add_argument('--gpu_count', type=int, default=1, help='Number of GPU per trial (default: %(default)s)')
-    parser.add_argument('--local_dir', default=os.getcwd(), help='Directory to save training results of tune (default: %(default)s)')
+    parser.add_argument('--cpu_count', type=int, default=4,
```
Comment (Eleven1Liu) [inline search_params.py:157]: As discussed, should we remove the new lines?

**Review by Eleven1Liu — CHANGES_REQUESTED** [review]

File: libmultilabel/model.py [hunk 6]
```diff
@@ -1,218 +1,112 @@
...
+                'pred_scores': torch.sigmoid(pred_logits).detach().cpu().numpy(),
+                'target': batch['labels'].detach().cpu().numpy()}
+
+    def validation_epoch_end(self, step_outputs):
+        eval_metric = MultiLabelMetrics(self.config)
+        for step_output in step_outputs:
+            eval_metric.add_values(y_pred=step_output['pred_scores'],
+                                   y_true=step_output['target'])
+        eval_metric.eval()
+        self.log_dict(eval_metric.get_metric_dict())
```
Comment (Eleven1Liu) [inline libmultilabel/model.py:72]: Set `use_cache` to False will run `eval()` before getting the metric dict. We add `use_cache` to calculate metrics in single run.
```
self.log_dict(eval_metric.get_metric_dict(use_cache=False))
```

**Review by Eleven1Liu — CHANGES_REQUESTED** [review]

File: libmultilabel/model.py [hunk 7]
```diff
@@ -1,218 +1,112 @@
...
-
-    def load_best(self):
-        best_ckpt_path = os.path.join(self.config.result_dir,
-                                      self.config.run_name, 'model_best.pt')
-        best_model = self.load(self.config, best_ckpt_path)
-        self.network = best_model.network
+        return optimizer
+
+    @abstractmethod
+    def shared_step(batch):
```
Comment (Eleven1Liu) [inline libmultilabel/model.py:52]:
```suggestion
    def shared_step(self, batch):
```

File: search_params.py (hunk 3 above)
Comment (henryyang42) [reply search_params.py:45]: The extension name doesn't matters here. Since we are saving and loading checkpoint here, `.ckpt` is more semantic meaningful for my taste. Also, PL tutorial use `.ckpt` throughout the guide.

File: search_params.py (hunk 4 above)
Comment (henryyang42) [reply search_params.py:26]: Merged

File: libmultilabel/model.py (hunk 6 above)
Comment (henryyang42) [reply libmultilabel/model.py:72]: Cool.

File: search_params.py (hunk 3 above)
Comment (Eleven1Liu) [reply search_params.py:45]: I have the same question after reading that guide. What would you suggest?

File: search_params.py (hunk 5 above)
Comment (henryyang42) [reply search_params.py:157]: Could we unify to 
```
    parser.add_argument('--vocab_file', type=str,
                        help='Path to a file holding vocabuaries (default: %(default)s)')
```
Making help on second line, which could be easier to read than single line?

**Comment by henryyang42** [comment]
@Eleven1Liu Are you interested in using seed_everything?
https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.utilities.seed.html#pytorch_lightning.utilities.seed.seed_everything

File: search_params.py (hunk 3 above)
Comment (henryyang42) [reply search_params.py:46]: I will keep it that way

**Comment by Eleven1Liu** [comment]
> @Eleven1Liu Are you interested in using seed_everything?
> https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.utilities.seed.html#pytorch_lightning.utilities.seed.seed_everything

Sounds great. Let's use `seed_everything` if the result are not affected.

File: libmultilabel/model.py (hunk 6 above)
Comment (Eleven1Liu) [reply libmultilabel/model.py:72]: What would the evaluation process of local_rank=1,2,3 like if there are 4 GPUs ?

File: search_params.py (hunk 5 above)
Comment (Eleven1Liu) [reply search_params.py:158]: Maybe we can follow the rules (see section indentation) here: https://www.python.org/dev/peps/pep-0008

**Review by sian-chen — CHANGES_REQUESTED** [review]

File: search_params.py (hunk 5 above)
Comment (henryyang42) [reply search_params.py:158]: Yes. But there are many different styles for [indentation](https://www.python.org/dev/peps/pep-0008/#indentation)
For the args we use this style for now.
```
# Aligned with opening delimiter.
foo = long_function_name(var_one, var_two,
                         var_three, var_four)
```

Let's discuss if we want to unify the style

File: libmultilabel/callbacks.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import shutil
+
+from pytorch_lightning.callbacks.early_stopping import EarlyStopping
+
+
+class EarlyStoppingWithCheckpoint(EarlyStopping):
```
[the remark this answers is not in the document]
Comment (henryyang42) [reply libmultilabel/callbacks.py:6]: Indeed

File: libmultilabel/model.py (hunk 6 above)
Comment (henryyang42) [reply libmultilabel/model.py:72]: If we are using `ddp`, each process will only use `sample size / GPU size` samples for evaluation (DistibutedRamdomSampler) by default in PL, the metrics may not be consistent across processes.  
We could override the sampler in dataloader only for dev/test loader to allow evaluation use whole dataset (https://github.com/ASUS-AICS/icd10-ai/blob/pl/icd/trainer.py)

File: libmultilabel/data_utils.py [hunk 9]
```diff
@@ -35,18 +35,16 @@ def __getitem__(self, index):
         data = self.data[index]
         return {
             'text': torch.LongTensor([self.word_dict[word] for word in data['text']][:self.max_seq_length]),
-            'label': torch.FloatTensor(self.label_binarizer.transform([data['label']])[0]),
-            'index': data.get('index', 0)
+            'labels': torch.FloatTensor(self.label_binarizer.transform([data['label']])[0]),
```
[the remark this answers is not in the document]
Comment (henryyang42) [reply libmultilabel/data_utils.py:38]: Sounds reasonable. Do we need to examine all variable names in this repo and remove `s` by this chance?

File: libmultilabel/callbacks.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (henryyang42) [reply libmultilabel/callbacks.py:6]: Another question is that in EarlyStoppingWithCheckpoint we only save model on rank 0 in ddp mode. I am not sure if thats trivial in ModelCheckpoint and EarlyStopping callback

File: libmultilabel/model.py (hunk 6 above)
Comment (Eleven1Liu) [reply libmultilabel/model.py:72]: Note: `use_cache` = `False` as default

File: libmultilabel/callbacks.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (henryyang42) [reply libmultilabel/callbacks.py:6]: After further investigating into their source code, we found that ``ModelCheckpoint`` will only save on rank zero using ``trainer.should_rank_save_checkpoint``. In such case, we could replace ``EarlyStoppingWithCheckpoint`` using a combination of ``ModelCheckpoint`` and ``EarlyStopping``.

**Comment by henryyang42** [comment]
I have fixed the issues we discussed earlier
- Rollback naming: label -> labels
- Replace EarlyStoppingWithCheckpoint with ModelCheckpoint and EarlyStopping
- Set use_cache=False in MultiLabelMetric

The change won't affect the reproducibility of rcv1/kim_cnn.
Can @Eleven1Liu please help validate on CAML and search_params.py? 
Thanks!

**Review by sian-chen — DISMISSED** [review]

**Comment by Eleven1Liu** [comment]
> I have fixed the issues we discussed earlier
> 
> * Rollback naming: label -> labels
> * Replace EarlyStoppingWithCheckpoint with ModelCheckpoint and EarlyStopping
> * Set use_cache=False in MultiLabelMetric
> 
> The change won't affect the reproducibility of rcv1/kim_cnn.
> Can @Eleven1Liu please help validate on CAML and search_params.py?
> Thanks!

@henryyang42  test completed. Everything's fine on  CAML.

**Review by Eleven1Liu — APPROVED** [review]

**Review by sian-chen — APPROVED** [review]
